Scaled tanh nonlinearity

f0k opened the pull request:
This extends PR #409 with better documentation (and correct markup) and slightly simplified code. If this is merged, it closes #409.
(I originally wanted to send @Botev a PR to his PR, but since no further work is required on top, I figured it would be easier to send a new PR to Lasagne instead.)

/edit: I also swapped the meaning of alpha and beta, since I found it extremely confusing that they were not matching the order of scale_in and scale_out in the constructor parameters. For a fact, Alexander got confused by this as well and reported LeCun's suggested factors in a different order than the other ones.

Botev commented:
Cool. Sorry for the confusion in the docs. Happy it finally made it in.

PS: Btw, I just noticed that the BN operator was merged in the main Theano branch Theano/Theano#3410 2 days ago. Maybe should have a formal BN layer now?

f0k commented:
> Maybe should have a formal BN layer now?

That one wasn't holding us back. It's mostly that there's no clean interface to include the running averages, but I guess we'll use my default_updates trick (https://gist.github.com/f0k/f1a6bd3c8585c400c190), I just have to do a PR.

skaae commented:
I tested your BN for reproducing the ladder network. It seems to work fine. The only thing i changed was an option to collect the statistics by running the entire dataset through the network in a single batch. This works fine for MNIST but will not work for larger datasets. The benefit is that I get the correct statistics.

f0k commented:
We should discuss that in #141, but...

> The only thing i changed was an option to collect the statistics by running the entire dataset through the network in a single batch. This works fine for MNIST but will not work for larger datasets.

That's good, it should be possible to turn this into a function that iterates over the dataset in mini-batches to collect those statistics. I've got something in place for my implementation, but results were worse than with the running averages collected during training, so I guess there's something wrong. If you get good statistics with a single-batch approach, we've got at least something to compare to.

f0k commented:
Back to topic, any comments on this PR? Is it good to merge?

Botev commented:
How would this be more beneficial in terms of computation (Running trough the mini batches for statistics). This would mean that either:
a) You run trough all mini batches to collect statistics, discarding all intermediate results (considering for any deep architecture). Thus this would be the equivalent of an extra full forward pass on the NN for each single mini batch.
b) You run trough all mini batches to collect statistics + you keep all intermediate results. Although this would be optimal in terms of speed the memory requirement will blow up too much for larger models. 
Am I missing something on how you guys collect these statistics (considering they change on every mini batch gradient update)?

f0k commented:
Our discussion was just about re-collecting the statistics over the full training set *after* training, instead of using the exponential moving averages collected *during* training to have something for the validation set. But we're off topic here, please have a look at #141 and post there if you have any questions / ideas!

f0k commented:
> Back to topic, any comments on this PR? Is it good to merge?

ebenolson commented on lasagne/nonlinearities.py:
```diff
@@ -60,6 +60,87 @@ def tanh(x):
 
 
+# scaled tanh
+class ScaledTanH(object):
```
Why `ScaledTanH` instead of `ScaledTanh`?

f0k replied:
It's camel case for Scaled Tangens Hyperbolicus, I assume. But I think `ScaledTanh` looks more natural... maybe we should support both?

Botev replied:
This I think is a matter of convention, but I just looked around in different tools and literature and I think @ebenolson is probably right and `ScaledTanh` is the much more widely used option.

ebenolson commented on lasagne/nonlinearities.py:
```diff
@@ -60,6 +60,87 @@ def tanh(x):
+
+    Parameters
+    ----------
+    scale_in : float32
```
why `float32` instead of `float`?

f0k replied:
I guess that was copy/pasted from the other docstrings here written by Martin Thoma. At some point, we need to figure out how to call different things and unify all the docstrings in the library accordingly. That's outside the scope of this PR.

ebenolson commented on lasagne/tests/test_nonlinearities.py:
```diff
@@ -22,20 +22,33 @@ def sigmoid(self, x):
+    def scaled_tanh(self, x):
+        return np.tanh(x)
+
+    def scaled_tanh_p(self, x):
```
what does `_p` stand for? parametrized?

f0k replied:
@Botev?

Botev replied:
The full thing should stand for `scaled_tanh_parametrized` as it is a test for the parametrized non standard `tanh`.

ebenolson commented on lasagne/nonlinearities.py:
```diff
@@ -60,6 +60,87 @@ def tanh(x):
+        self.scale_out = scale_out
+
+    def __call__(self, x):
+        if self.scale_in != 1:
```
how much does this improve the efficiency vs just

    return self.scale_out * theano.tensor.tanh(x * self.scale_in)

f0k replied:
Not at all, it's optimized away:
```
import theano
>>> x = theano.tensor.matrix('x')
>>> x.dtype
float32
>>> (1.0*x*1.0).dtype
float32
>>> theano.printing.debugprint(1.0*x*1.0)
Elemwise{mul,no_inplace} [@A] ''   
 |Elemwise{mul,no_inplace} [@B] ''   
 | |DimShuffle{x,x} [@C] ''   
 | | |TensorConstant{1.0} [@D]
 | |<TensorType(float32, matrix)> [@E]
 |DimShuffle{x,x} [@F] ''   
   |TensorConstant{1.0} [@D]
>>> theano.printing.debugprint(theano.function([x], 1.0*x*1.0))
DeepCopyOp [@A] ''   0
 |<TensorType(float32, matrix)> [@B]
```
I'll replace it by the one-liner.

Botev replied:
Ah sorry for that it was written in my first days in using Lasagne/Theano where I did not know what is happening.

ebenolson commented:
couple of nitpicks

f0k commented:
Changed implementation to one-liner (https://github.com/Lasagne/Lasagne/pull/430#discussion_r40545331), added `ScaledTanh` alias for people reading it as `scaled tanh()` instead of `scaled tangens hyperbolicus` -- both are valid interpretations, I guess.
Will leave the test and the docstring argument type unchanged, but I'm always happy about nitpicking! :)
